pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionAssignmentTailer.java

```diff
@@ -25,95 +25,157 @@
+    public synchronized void close() {
+        log.info("Closing function assignment tailer");
         try {
             isRunning = false;
-            if (tailerThread != null && tailerThread.isAlive()) {
-                tailerThread.interrupt();
+
+            if (tailerThread != null) {
+                while (true) {
+                    tailerThread.interrupt();
+
+                    try {
+                        tailerThread.join(5000, 0);
+                    } catch (InterruptedException e) {
+                        log.warn("Waiting for assignment tailer thread to stop is interrupted", e);
+                    }
+
+                    if (tailerThread.isAlive()) {
+                        log.warn("Assignment tailer thread is still alive.  Will attempt to interrupt again.");
+                    } else {
+                        break;
+                    }
+                }
+                tailerThread = null;
             }
             if (reader != null) {
                 reader.close();
+                reader = null;
             }
+
+            hasExited = null;
+            exitOnEndOfTopic = false;
+            
         } catch (IOException e) {
             log.error("Failed to stop function assignment tailer", e);
         }
-        log.info("Stopped function assignment tailer");
+    }
+    
+    private Reader<byte[]> createReader(MessageId startMessageId) throws PulsarClientException {
+        log.info("Assignment tailer will start reading from message id {}", startMessageId);
+
+        return readerBuilder
+                .subscriptionRolePrefix(workerConfig.getWorkerId() + "-function-assignment-tailer")
+                .readerName(workerConfig.getWorkerId() + "-function-assignment-tailer")
+                .topic(workerConfig.getFunctionAssignmentTopic())
+                .readCompacted(true)
+                .startMessageId(startMessageId)
+                .create();
     }
 
-    public void processAssignment(Message<byte[]> msg) {
-
-        if(msg.getData()==null || (msg.getData().length==0)) {
-            log.info("Received assignment delete: {}", msg.getKey());
-            this.functionRuntimeManager.deleteAssignment(msg.getKey());
-        } else {
-            Assignment assignment;
-            try {
-                assignment = Assignment.parseFrom(msg.getData());
-            } catch (IOException e) {
-                log.error("[{}] Received bad assignment update at message {}", reader.getTopic(), msg.getMessageId(), e);
-                throw new RuntimeException(e);
+    private Thread getTailerThread() {
+        Thread t = new Thread(() -> {
+            while (isRunning) {
+                try {
+                    Message<byte[]> msg = reader.readNext(5, TimeUnit.SECONDS);
+                    if (msg == null) {
+                        if (exitOnEndOfTopic && !reader.hasMessageAvailable()) {
+                            break;
+                        }
```

Comment:
is it simpler if we do
while(isRunning) {
if (exitOnEndOfTopic && !available) break;
try { read message... }

Reply:
It's safer to wait for a timeout period to make sure no messages just arrived late